### DISCUSSION TITLE
Packages For Node Project

### DISCUSSION THREAD

**Discussion by momin-riyadh** [momin-riyadh/fullstack#6] [Guideline]
In day-to-day NodeJS, you will probably use

1. Some API frameworks, e.g., Express or NestJS

2. Some databases with ORM using either Sequelize for MySQL/Postgres or Mongoose for MongoDB

3. Publish/subscribe to message queues like rabbitmq or streams like Kafka

4. Work with cache like Redis

5. Handle integrations with third-party APIs (Axios, fetch, etc., HTTP client. Need good command of JSON API calls and response handling)

6. Background jobs/cron jobs with factory or bull

7. Real-time peer-to-peer connections with socket.io

8. Unit testing or integration testing with jest/superset

9. Logging with Winston / Pino / Morgan (use morgan-body if you want to log requests/responses, useful for debugging in staging and dev environments)

10. Authentication with passport.js, or custom authentication using cookie-parser, jsonwebtoken, or OAuth
11. Templating Engine EJS, Pug, Marko

These are things used in standard production NodeJS codebases. However, to be a good backend developer in NodeJS, knowing these will help you.

1. Good use of JavaScript/Typescript data structures

2. Good command of the language, e.g., syntax and good practices

3. Asynchronous programming - using async await/promises efficiently
4. Debugging skills - being able to read error messages and figure out what is going wrong with your (or someone else's code) - this is general advice

5. Knowing how to write practical unit tests

6. Last but not least, having the curiosity to find out how things work. You can review the source code using an open-source library and see how it was implemented. You can also learn a lot from reading other people's code.

**Comment by chayansarker** [comment]
Please suggest ORM for MySql

**Reply by momin-riyadh** [reply]
https://sequelize.org/
